tensorflow-ops/src/TensorFlow/Gradient.hs

```diff
@@ -165,6 +166,10 @@ gradients y xs = build $ do
         (\f x -> fromMaybe (error $ "no NodeDef found for " ++ show x) (f x))
         . flip Map.lookup
     let (gr, nodeMap) = createGraph yName nodeDefLookup
+        xnodes = mapMaybe (\x -> nodeMap ^. (at . outputNodeName . renderedOutput $ x)) xs
+        -- make a set of the nodes reachable from the xnodes (lacking the xnodes themselves)
+        reachableSet = (computeReachableSet xnodes gr) `IntSet.difference` (IntSet.fromList xnodes)
```

Comment:
Sorry, I missed an edge case when I made this suggestion. If one member of `xnodes` is reachable from another member of `xnodes`, then we need to keep it in this set.

Maybe one solution is to remove the root of each `Tree` returned by the `FGL.dff` call before calling `FGL.preorderF`.

Reply:
Good catch! A test for this case and your suggestion has been implemented.